### PR TITLE
Fix MSRVs for standalone crates

### PR DESCRIPTION
**Pull request by BenjaminBrienen**
# Objective

MSRV in the standalone crates should be accurate

## Solution

Determine the msrv of each crate and set it

## Testing

Adding better msrv checks to the CI is a next-step.

### CODE REVIEW

**Comment by mockersf** [comment]
I think most of those changes are not justified outside of "you think it looks better"

I would prefer to not change the version of dependencies, even for something compatible, without a need to change it for that dependency. Even worse to do it for everything at once.

For MSRV, adding it to most crates will just make it more painful for us to maintain: for the crates that can't be used outside of Bevy, adding it doesn't bring us any value and will just need more work.

File: errors/Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,7 @@
 [package]
 name = "errors"
 edition = "2021"
+version = "0.1.0"
 description = "Documentation and tests for Bevy's error codes"
 publish = false
 license = "MIT OR Apache-2.0"
```
Comment (BenjaminBrienen) [inline errors/Cargo.toml:7]: Missing

File: benches/Cargo.toml [hunk 2]
```diff
@@ -4,12 +4,13 @@ edition = "2021"
...
 publish = false
 license = "MIT OR Apache-2.0"
+rust-version = "1.78.0"
 
 [dev-dependencies]
 glam = "0.29"
 rand = "0.8"
 rand_chacha = "0.3"
-criterion = { version = "0.3", features = ["html_reports"] }
+criterion = { version = "0.5", features = ["html_reports"] }
```
Comment (BenjaminBrienen) [inline benches/Cargo.toml:13]: Undo

File: crates/bevy_core_pipeline/src/smaa/mod.rs [hunk 3]
```diff
@@ -35,6 +35,8 @@ use crate::{
     core_3d::graph::{Core3d, Node3d},
 };
 use bevy_app::{App, Plugin};
+#[cfg(not(feature = "smaa_luts"))]
```
Comment (BenjaminBrienen) [inline crates/bevy_core_pipeline/src/smaa/mod.rs:38]: Move to different PR

File: .gitignore [hunk 4]
```diff
@@ -1,18 +1,18 @@
 # Rust build artifacts
```
Comment (BenjaminBrienen) [inline .gitignore:1]: Move to different PR

File: crates/bevy_ecs/compile_fail/Cargo.toml [hunk 5]
```diff
@@ -6,6 +6,7 @@ homepage = "https://bevyengine.org"
 repository = "https://github.com/bevyengine/bevy"
 license = "MIT OR Apache-2.0"
 publish = false
+rust-version = "1.81.0"
```
Comment (BenjaminBrienen) [inline crates/bevy_ecs/compile_fail/Cargo.toml:9]: Undo

File: output.jsonc
Comment (BenjaminBrienen) [inline output.jsonc:1]: Undo

File: tools/build-templated-pages/Cargo.toml [hunk 6]
```diff
@@ -1,18 +1,18 @@
 [package]
 name = "build-templated-pages"
 edition = "2021"
+version = "0.1.0"
 description = "Tool that checks and fixes undocumented features and examples"
 publish = false
 license = "MIT OR Apache-2.0"
+rust-version = "1.73.0"
```
Comment (BenjaminBrienen) [inline tools/build-templated-pages/Cargo.toml:8]: Undo

File: tools/build-wasm-example/Cargo.toml [hunk 7]
```diff
@@ -1,14 +1,15 @@
 [package]
 name = "build-wasm-example"
 edition = "2021"
+version = "0.1.0"
 description = "Tool for building example for Wasm"
 publish = false
 license = "MIT OR Apache-2.0"
-
+rust-version = "1.74.1"
```
Comment (BenjaminBrienen) [inline tools/build-wasm-example/Cargo.toml:8]: Undo

File: tools/ci/Cargo.toml [hunk 8]
```diff
@@ -1,14 +1,16 @@
 [package]
 name = "ci"
 edition = "2021"
+version = "0.1.0"
 description = "Tool that enables running CI checks locally."
 publish = false
 license = "MIT OR Apache-2.0"
+rust-version = "1.72.1"
```
Comment (BenjaminBrienen) [inline tools/ci/Cargo.toml:8]: Undo

File: tools/compile_fail_utils/Cargo.toml [hunk 9]
```diff
@@ -6,9 +6,10 @@ homepage = "https://bevyengine.org"
 repository = "https://github.com/bevyengine/bevy"
 license = "MIT OR Apache-2.0"
 publish = false
+rust-version = "1.78.0"
```
Comment (BenjaminBrienen) [inline tools/compile_fail_utils/Cargo.toml:9]: Undo

File: tools/example-showcase/Cargo.toml [hunk 10]
```diff
@@ -1,19 +1,19 @@
 [package]
 name = "example-showcase"
 edition = "2021"
+version = "0.1.0"
 description = "Tool for running examples or generating a showcase page for the website."
 publish = false
 license = "MIT OR Apache-2.0"
+rust-version = "1.74.1"
```
Comment (BenjaminBrienen) [inline tools/example-showcase/Cargo.toml:8]: Undo

**Comment by BenjaminBrienen** [comment]
image

**Comment by BenjaminBrienen** [comment]
To address the other feedback:

- I do think it looks better. The alternative option would be keeping the version numbers in-line with what versions are actually being used, which is incredibly noisy and unnecessary.

- I have ensured that no dependencies have been downgraded or updated by `diff`ing `Cargo.lock`.

- I have made sure that the `rust-version` is only specified for public crates. The maintenance cost will be hopefully the exact same as it is now or even less after I have finished improving the CI. We can make it so that the developer doesn't have to edit the value themselves, even. The bot can do that.

**Comment by mockersf** [comment]
> * I do think it looks better. 

I don't. And I especially dislike doing it in a change across all crates, without a reason to do it related to the dependency.

> * I have made sure that the `rust-version` is only specified for public crates. The maintenance cost will be hopefully the exact same as it is now or even less after I have finished improving the CI. We can make it so that the developer doesn't have to edit the value themselves, even. The bot can do that.

I don't think having separate msrv for crates that are not standalone has enough value to bother, I don't think they should be added and I won't support adding those checks in CI. This can be revisited later for a 1.0 release, but for the moment the Bevy msrv policy is "latest stable", let's not go into details crate per crate.

**Comment by BenjaminBrienen** [comment]
Ok. I'd like to discuss the second point a bit more if you don't mind. I thought the design philosophy of bevy is to have each crate be usable on their own. A project could just use `bevy_ecs`, for example. Why wouldn't we want the usability there to be the same as the main `bevy` crate? This is assuming we set the msrv of `bevy` accurately for usability and to provide value to dependents. To me, it works the same way for each public crate. Can you please tell me your perspective on that aspect?

File: crates/bevy_image/Cargo.toml [hunk 11]
```diff
@@ -44,26 +45,26 @@ bevy_reflect = { path = "../bevy_reflect", version = "0.15.0-dev", features = [
 bevy_utils = { path = "../bevy_utils", version = "0.15.0-dev" }
 
 # rendering
-image = { version = "0.25.2", default-features = false }
+image = { version = "0.25", default-features = false }
```
Comment (rparrett) [inline crates/bevy_image/Cargo.toml:48]: These aren't equivalent, and we want at least `0.25.2`. See https://github.com/bevyengine/bevy/pull/14505.

I think there are other instances where this is known to be true, and probably more yet that could be lurking issues if this were merged.

Comment (BenjaminBrienen) [reply crates/bevy_image/Cargo.toml:48]: Wouldn't a `cargo update` fix that? If the developer doesn't want to do it manually, a post-pull hook would solve that problem.

Comment (rparrett) [reply crates/bevy_image/Cargo.toml:48]: What would be the benefit to specifying a version requirement allowing versions that are known to cause issues?

**Comment by alice-i-cecile** [comment]
> Ok. I'd like to discuss the second point a bit more if you don't mind. I thought the design philosophy of bevy is to have each crate be usable on their own. A project could just use `bevy_ecs`, for example. Why wouldn't we want the usability there to be the same as the main `bevy` crate? This is assuming we set the msrv of `bevy` accurately for usability and to provide value to dependents. To me, it works the same way for each public crate. Can you please tell me your perspective on that aspect?

So, this ties into https://github.com/bevyengine/bevy/issues/16172. Only the standalone crates (bevy_ecs, bevy_math, bevy_color, bevy_ptr, bevy_reflect) should have their own MSRV. For any crate that relies on bevy_ecs, their MSRV should be fully tied to bevy_ecs, which for the foreseeable future is going to be on "latest stable".

**Comment by BenjaminBrienen** [comment]
I'll update this PR with the feedback when I have some free time. I'll fix some of the minor versions and remove more of the msrvs as requested. The main goal is consistency, correctness, and usability, so I'm glad to work with the feedback on that.

**Comment by BenjaminBrienen** [comment]
Marking as blessed because I reduced the scope to exactly the changes endorsed by @alice-i-cecile.

**Review by alice-i-cecile — APPROVED** [review]
I agree with this scope and these fixes :)

**Review by mockersf — APPROVED** [review]
